pallets/nodle-staking/src/migrations.rs

```diff
@@ -0,0 +1,297 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+use super::{BalanceOf, Config, Validator};
+
+use frame_support::{
+    pallet_prelude::*,
+    traits::StorageInstance,
+    weights::{constants::RocksDbWeight, Weight},
+};
+
+use sp_runtime::traits::Zero;
+use sp_std::vec::Vec;
+
+/// This migration is in charge of moving the validators from POA to Staking pallet to
+/// work with the last chain reboot.
+/// We have to do two things:
+/// - read "Validators" storage instance from POA.
+/// - with read info update the "Invulnerables" & "ValidatorState" of staking pallet.
+#[allow(dead_code)]
+pub fn poa_validators_migration<T: Config>() -> Weight {
+    // Buffer to account for misc checks
+    let mut weight: Weight = 1_000;
+
+    log::info!("🕊️ Starting Poa validators migration...");
+
+    struct POAValidatorsStorageValuePrefix;
+    impl StorageInstance for POAValidatorsStorageValuePrefix {
+        const STORAGE_PREFIX: &'static str = "Validators";
+        fn pallet_prefix() -> &'static str {
+            "Poa"
+        }
+    }
+
+    struct StakingValidatorStateStorageMapPrefix;
+    impl StorageInstance for StakingValidatorStateStorageMapPrefix {
+        const STORAGE_PREFIX: &'static str = "ValidatorState";
+        fn pallet_prefix() -> &'static str {
+            "NodleStaking"
+        }
+    }
+
+    struct StakingInvulnerablesStorageValuePrefix;
+    impl StorageInstance for StakingInvulnerablesStorageValuePrefix {
+        const STORAGE_PREFIX: &'static str = "Invulnerables";
+        fn pallet_prefix() -> &'static str {
+            "NodleStaking"
+        }
+    }
+
+    #[allow(type_alias_bounds)]
+    type POAValidators<T: Config> =
+        StorageValue<POAValidatorsStorageValuePrefix, Vec<T::AccountId>, OptionQuery>;
+
+    #[allow(type_alias_bounds)]
+    type StakingInvulnerables<T: Config> =
+        StorageValue<StakingInvulnerablesStorageValuePrefix, Vec<T::AccountId>, OptionQuery>;
+
+    #[allow(type_alias_bounds)]
+    type StakingValidatorState<T: Config> = StorageMap<
+        StakingValidatorStateStorageMapPrefix,
+        Twox64Concat,
+        T::AccountId,
+        Validator<T::AccountId, BalanceOf<T>>,
+        OptionQuery,
+    >;
+
+    log::info!(
+        "<POAValidators<T>>::exits()? {:?}",
+        <POAValidators<T>>::exists()
+    );
+    log::info!(
+        "<StakingInvulnerables<T>>::exits()? {:?}",
+        <StakingInvulnerables<T>>::exists()
+    );
+
+    if !<POAValidators<T>>::exists() {
```

Comment:
It would be safe to assume that any call to `exists` count as an additional read weight we need to account for

Reply:
:+1:  Done